esp/esdllib/esdl_script.cpp

```diff
@@ -53,26 +53,65 @@ class CEsdlScriptContext : public CInterfaceOf<IEsdlScriptContext>
     virtual IEsdlFunctionRegister* queryFunctionRegister() const override { return functionRegister; }
     virtual void setTraceToStdout(bool val) override
     {
-        traceToStdout = val;
+        if (val != traceToStdout)
+        {
+            traceToStdout = val;
+            tracer->setSink(val ? consoleSink.get() : jlogSink.get());
+        }
     }
     virtual bool getTraceToStdout() const override { return traceToStdout; }
     virtual void setTestMode(bool val) override { testMode = val; }
     virtual bool getTestMode() const override { return testMode; }
+    virtual ITracer& trace() const override { return *tracer; }
+    virtual bool enableMasking(const char* domainId, uint8_t version) override
+    {
+        if (!maskingEngine)
+        {
+            trace().uwarnlog("enable masking request not completed - no masking engine");
+            return false;
+        }
+        if (maskingEnabled())
+        {
+            trace().uwarnlog("enable masking request not completed - already enabled");
+            return false;
+        }
+        Owned<IDataMaskingProfileContext> ctx(maskingEngine->getContext(domainId, version, tracer));
+        if (!ctx)
+        {
+            trace().uwarnlog("enable masking request not completed - no context for '%s:%hhu'", domainId, version);
+            return false;
+        }
+        maskerStack.emplace_back(ctx.getClear());
+        return maskingEnabled();
+    }
+    virtual bool maskingEnabled() const override { return !maskerStack.empty(); }
+    virtual IDataMaskingProfileContext* masker() const override { return maskerStack.empty() ? nullptr : maskerStack.back().get(); }
 private:
     Owned<IEspContext>            espCtx;
     IEsdlFunctionRegister*        functionRegister = nullptr;
     Owned<ISectionalXmlDocModel>  docModel;
     bool                          traceToStdout = false;
     bool                          testMode = false;
+    Owned<CModularTracer>         tracer;
+    Owned<IModularTraceMsgSink>   jlogSink;
+    Owned<IModularTraceMsgSink>   consoleSink;
+    Owned<IDataMaskingEngine>     maskingEngine;
+    using MaskerStack = std::list<Owned<IDataMaskingProfileContext> >;
+    MaskerStack                   maskerStack;
 public:
-    CEsdlScriptContext(IEspContext* _espCtx, IEsdlFunctionRegister* _functionRegister)
+    CEsdlScriptContext(IEspContext* _espCtx, IEsdlFunctionRegister* _functionRegister, IDataMaskingEngine* _engine)
         : functionRegister(_functionRegister)
     {
         if (_espCtx)
             espCtx.set(_espCtx);
         else
             espCtx.setown(createEspContext(nullptr));
         docModel.setown(createSectionalXmlDocModel(this));
+        tracer.setown(new CModularTracer());
+        jlogSink.setown(tracer->getSink());
+        consoleSink.setown(new CConsoleTraceMsgSink());
+        tracer->setSink(jlogSink);
```

Comment:
Why get and then set the default sink to the same value without any change?

Reply:
If it is necessary, could move this up a line to group the two calls together with a comment explaining why necessary.

Reply:
If I had to guess, `tracer` did not originally create the default sink. Both sinks were created here, and one was applied as the default. The creation of the default was updated, but setting the default wasn't.